Notify client close

viktordick opened the pull request:
Implementation of the idea discussed in #308. The tests don't pass yet and I should also implement additional tests for this, but hopefully this helps with the discussion.

viktordick commented:
@bertjwregeer Is there a chance that this PR will be accepted if I fix the tests and add new ones for the new feature? Or do you think our request is too exotic and the added complexity should be avoided?

bertjwregeer commented on src/waitress/channel.py:
```diff
@@ -194,7 +226,7 @@ def received(self, data):
 
         if requests:
-            self.requests = requests
+            self.requests.extend(requests)
```
This becomes very unfair, if someone were to keep pipelining requests, the task thread that is servicing the requests will get stuck inside it's while self.requests loop with the task lock held... meaning that someone pipelining one request after the other could stop any other requests from moving forward because they would never get scheduled.

viktordick replied:
If I understand the code correctly, the task lock is defined for each channel and each channel is only responsible for one socket, so different connections should not influence each other. If one client keeps pipelining requests, `len(self.requests)` will become larger than `max_requests_per_channel` and no more `read` is performed until enough requests have been answered.

Is this any different from the behavior before if a client kept writing more and more requests into the same connection?

bertjwregeer replied:
On the line right after this you re-add the task to the queue of items to be processed. This means that if thread 1 is already processing all requests (and thus has the task lock), another thread may pick up that it needs to work on this channel because it was added to the queue and it will wait until it can get the task lock.

For each request that is added it will block. Each time a request is popped off, dropping below the maximum another request may be read.

bertjwregeer replied:
Each time you call `self.server.add_task(self)` it adds the current channel to the list of items to be processed.

However only a single thread can hold the task lock. If we keep adding to `self.requests` that single thread that is processing the current request while we read this one will continue to process requests.

In the mean time each time we added ourselves to the backlog on line 216 we could potentially get scheduled and block waiting for us to get the task lock.

You'd need to make sure you only schedule a task if its not already running. You'd want to do a non-blocking acquire, and only if you acquired the lock should you add yourself to the task queue... although I foresee that if multiple reads are allowed but the threads are really busy you'd still potentially schedule the same channel multiple times, which could still lead to the same deadlock.

bertjwregeer replied:
Instead we'd probably want to rewrite waitress so that if there are multiple requests, the channel reschedules itself for execution at the end of the line, rather than trying to process all of the requests in the queue. This way pipelined requests are scheduled more fairly.

All in all that is a much larger change, since requests have to be processed in the order received because the client expects responses in the order it sent them.

viktordick replied:
Right, thanks, now I understand. Only one thread will ever execute for a given channel due to the `task_lock`, but it might happen that one channel blocks all threads that `waitress` has available for tasks to be finished, while only one of them does any actual work.

So yes, it seems that `service` should only `pop` one request to work on it and, just before finishing, call `add_task` again if there are more requests. I will try to think about how to accomplish this without race conditions.

mmerickel replied:
I believe this has been resolved now, as the while loop as been removed and `channel.service()` only handles one request at a time, ensuring to invoke `server.add_task(self)` if there are more requests to execute keeping things fair.

viktordick commented:
@bertjwregeer I changed the implementation so `service()` only processes one request and re-inserts itself using `add_task` if there are more requests. This requires a lock to cleanly communicate if the main thread should call `add_task` or if the task thread should do it and prevent race conditions. However, I am pretty sure that I could re-use the `task_lock` for this. It is no longer necessary to guard the complete `service` method with this lock since by only re-inserting the next task at the very end of the function, it is ensured that no two requests of the same channel are processed in the wrong order or in parallel.

However, this makes the difference to `master` rather large since most of the content in `service()` changed the indentation level.

One change I am not sure about is the handling of the output buffers. Due to the change, they are flushed after each request separately. Before, if the client sent multiple short requests in a pipeline such that they would be picked up with a single `read()`, the output buffers might only be flushed once all of them were processed (I think).

I did some obvious adjustments to the tests. I also dropped `test_service_no_requests` since it does not work anymore (`service()` assumes that there is a request to be processed), but it anyhow does not handle any real life situation, `add_task` is only called if there is a request to be processed.

There are two more failing tests that I will check next. And, of course, I need to add tests for the new functionality.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -203,9 +222,12 @@ def received(self, data):
-            self.requests = requests
-            self.server.add_task(self)
+        with self.task_lock:
+            if requests:
```
the local requests var can be tested prior to acquiring the lock

mmerickel commented on src/waitress/channel.py:
```diff
@@ -203,9 +222,12 @@ def received(self, data):
+                self.requests.extend(requests)
+                if not self.task_running:
+                    self.server.add_task(self)
+                    self.task_running = True
```
isn't task_running redundant? if there's an item in `self.requests` then a task should be running, and we ensure we don't pop each request until a task is complete

viktordick replied:
This was necessary because `self.requests` might drop to an empty list at the beginning of `service`. But I guess by postponing `self.requests.pop(0)` until later (in the block at the end of `service` inside the `task_lock`), it is not necessary and testing whether `self.requests` is empty or not is sufficient.

mmerickel commented on src/waitress/task.py:
```diff
@@ -563,6 +563,10 @@ def get_environment(self):
 
+        check = getattr(self.channel, 'check_client_disconnected', None)
+        if callable(check):
+            environ['check_client_disconnected'] = check
```
when would this not be a callable?

viktordick replied:
It was only necessary in one of the tests that used a `DummyChannel`. But I guess it is better to extend that class accordingly.

mmerickel replied:
Ah, for the tests - yeah I'd prefer to fix the tests.

The environ var is going to need to be namespaced to waitress for now. I'd call it `environ['waitress.is_connected']` or something?

mmerickel commented on src/waitress/channel.py:
```diff
@@ -44,10 +44,13 @@ class HTTPChannel(wasyncore.dispatcher):
     sent_continue = False  # used as a latch after sending 100 continue
     total_outbufs_len = 0  # total bytes ready to send
     current_outbuf_count = 0  # total bytes written to current outbuf
+    client_disconnected = False
```
how is this variable different from `self.connected` which is used to raise `ClientDisconnected` exceptions?

mmerickel replied:
In general I'm concerned that the code hasn't been updated with any real coordination around `close_when_flushed`, `client_disconnected`, `connected`, `will_close`, etc. I'd like a sense for when you think each one is currently being used to justify adding a new variable.

viktordick replied:
Thanks, I will have to think a bit more about these. When starting with this, I was hoping not to have to deal with the response end of the request processing, but I guess that was too optimistic. Is there any documentation that could help me besides reading the source code and maybe asking questions?

viktordick replied:
Hm if I use `not self.connected` instead of `self.client_disconnected`, the tests still all succeed. But from what I can see in the code, it seems to me that this might prevent some cleanup (like the calls to `request.close()` in `service()` and the flushing of the output buffers in `handle_close()`).
@mmerickel or @bertjwregeer, can you confirm or alleviate my suspicion?

bertjwregeer commented:
@viktordick thanks for updating it, I haven't taken a look yet but I saw that @mmerickel provided some feedback. I've been busy moving, hopefully once things things calm down a little bit I'll get back to this and help figure out a best path forward.

mmerickel commented:
Thanks @viktordick for going through this, I think I agree with most of your analysis. A channel is never reused so there is no concern about the buffer cleanup or `connected` getting reset to `True`.

From what I'm seeing, `connected` is the right hook to use here to indicate that the connection is dead and the client is gone. This is what your current code is using. It sounds like this is a concern for you but I'm not clear why or what you think another variable will help capture better? Is there a lifecycle that you want to try to catch before it is closed?

I noticed that there's a potential race condition in `channel.received` where we are using `_flush_some` to write out a response. The underlying assumption in that block of code is that if we're reading, then no requests are active, but of course this PR breaks that assumption.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -153,6 +161,9 @@ def handle_read(self):
             self.received(data)
+        else:
+            # Client disconnected.
+            self.connected = False
```
I'm struggling to understand why this is being added / why it's necessary now versus before.

viktordick replied:
The idea of this PR is to detect a client disconnect even before the application starts to write the response, so the application may check at some strategic points if the client is still interested in the answer.

In a `select` loop, a socket that has been closed by the other side makes itself known by being readable (it is included in the readable list of the result of `select`), but if we read something, we get an empty answer. Therefore, we keep the channel `readable` (at least if there are not too many requests already queued from the channel) and if `handle_read` is then called and the read yields an empty response, this means the client has disconnected.

viktordick commented:
> Thanks @viktordick for going through this, I think I agree with most of your analysis. A channel is never reused so there is no concern about the buffer cleanup or `connected` getting reset to `True`.
But why do we then need `close_when_flushed` and `will_close`? `ClientDisconnected` is raised if the task tries to write part of the response into the output buffers and the client has disconnected. But why is `service` then going through all the steps instead of simply setting `connected` to `False`, emptying `self.requests` and returning? Even if `handle_close` should be called from the main thread instead of the task thread (which is the case I guess), this only explains the existence of `will_close`, not of `close_when_flushed` - I think.

I could imagine some sort of graceful shutdown which might use `close_when_flushed` to process all currently running tasks, but not accept any more. However, I do not see any code that might implement something like this.

> I noticed that there's a potential race condition in `channel.received` where we are using `_flush_some` to write out a response. The underlying assumption in that block of code is that if we're reading, then no requests are active, but of course this PR breaks that assumption.
Thanks, I will check this out

mmerickel commented:
> But why do we then need close_when_flushed and will_close

So there's a bunch of reasons to close via client, wsgi app, or process:

- client not using keep-alive (`close_when_flushed = True`)
- unhandled wsgi app exception (`close_when_flushed = True`)
- client unexpected hangup (`connected = False`)
- graceful server-triggered shutdown (not well supported but triggered via `Channel.cancel` / `will_close = true`)

> I could imagine some sort of graceful shutdown which might use close_when_flushed to process all currently running tasks, but not accept any more. However, I do not see any code that might implement something like this.

This is sort of supported right now via the `Channel.cancel` method which is invoked from the threadpool (TaskDispatcher). However it might not be as graceful as you'd expect. Right now it stops ASAP versus after draining the pending requests. I think as we get further into supporting a real graceful shutdown that behavior would change.

mmerickel commented:
Also yes, `will_close` is being used to ensure we are invoking `handle_close` from within the event loop and not from the channel's thread. Like I said above I think that its meaning would change if we added a better graceful shutdown. Likely to something closer to "close after flushing the current list of requests plus one we might be reading right now, and do not start reading any new ones".

bertjwregeer commented on src/waitress/channel.py:
```diff
@@ -360,88 +377,93 @@ def _flush_outbufs_below_high_watermark(self):
+                else:
+                    # Give responsibility to add the next task back to the main
+                    # thread
+                    self.task_running = False
```
This is not used anywhere else in this PR.

viktordick replied:
Oops, this was a leftover from an earlier version. Thanks for finding it.

bertjwregeer commented on tests/test_task.py:
```diff
@@ -839,6 +840,34 @@ def test_get_environment_values(self):
         self.assertEqual(environ["wsgi.input_terminated"], True)
         self.assertEqual(inst.environ, environ)
 
+    def test_client_disconnect(self):
```
Is this test incomplete on purpose?

viktordick replied:
Also a leftover. I first tried to implement the test in `test_task`, but later found it easier to implement it in `test_channel`.

bertjwregeer reviewed (commented):
I like where this is going. More graceful shutdown would be fantastic, and its something that I would like to add, so keeping that in mind while continuing down this path would be great.

viktordick commented:
I secured the writing into the `outbufs` in `reeceived` with the `outbuf_lock`, which seems to work well.

I'd say this PR is ready for a re-review. One check seems to be stuck - I had this problem with some other PR on github once and it seemed to be necessary to close and re-open the PR to get the tests to run again. Should I do this here also?

bertjwregeer commented:
Maintainers have the ability to hit the re-run all jobs button. So no need to close/open this PR.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -176,15 +186,14 @@ def received(self, data):
+                        self.current_outbuf_count += num_bytes
+                        self.total_outbufs_len += num_bytes
+                        self.sent_continue = True
+                        self._flush_some()
```
I think the problem with this is that we do not want to send responses out of order. Now that we can be reading a new request while processing a current request - this needs to be done in such a way that it does not send data on the outbufs until the current request is completed with its processing.

This may require checking if a current request is executing, and if so then buffering this data somewhere and flushing it when the request is complete. Or maybe more simply this entire branch of logic that handles 100 Continue needs to be moved into a task itself instead of being handled here.

I think the problem I'm identifying here is real but I'm not confident in my proposed solutions.

mmerickel replied:
Simplest solution may be to stop processing lookahead requests and don't resume until the current request is completed when we hit an expect-continue.

viktordick replied:
A very good point. This work really allows me to understand more about the HTTP protocol than I ever thought I would need. I was not aware of the meaning and mechanics of `100 Continue` before.

I guess I will first add a test that pushes a request that waits a second before returning a defined result and immediately afterwards pushes a request with expect-continue, checking that the responses come in the right order (which will fail now).

Your suggested solution should work, although I think it would hinder the detection of client disconnects if an expect-continue occurs. Maybe I can think of something better, else this solution will have to do.

viktordick replied:
I added a (currently failing) test to check for this. The thing is - I am pretty sure the situation is not covered on the current master branch either. At least not if `handle_read` finds one regular request and one with expect-continue that are both read with the same `recv` call. But now that we have identified the potential problem, we should fix it.

mmerickel replied:
hmm you may be right that it's broken right now if it receives one full request followed by an expect request in the same recv call.

It's worth noting also that waitress is caching `self.sent_continue` for the lifetime of the entire connection, so there's an expectation that it won't receive more than one expect request on the same connection and I'm not sure how valid that assumption is.

bertjwregeer replied:
It's an invalid assumption according to the RFC's, but no web browser (and I would argue almost all clients) don't actually support sending Expect Continue's in the first place, which is why nobody has yelled about it before.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -360,88 +375,89 @@ def _flush_outbufs_below_high_watermark(self):
-            while self.requests:
-                request = self.requests[0]
+        """Execute one request. If there are more, we add another task to the
+        server at the end."""
```
This is a noop comment, but just wanted to note that I like this workflow change as it prevents a connection from dominating a thread and tasks will be more fairly queued across connections.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -360,88 +375,89 @@ def _flush_outbufs_below_high_watermark(self):
             self.server.pull_trigger()
+            with self.task_lock:
+                if self.requests:
+                    self.requests.pop(0)
```
this first `if self.requests` shouldn't ever be false, right?

viktordick replied:
If `task.close_on_finish` is set, `self.requests` is emptied in line 429.

I could move the whole `if self.connected` block into the `else` block of `if task.close_on_finish`, except for the `pull_trigger`, which would then occur outside the block and after `add_task`. Not sure if that is a problem.

mmerickel commented:
I had only a couple items to note in my review - this is looking great and getting really close.

viktordick commented:
I finished implementing a `send_continue` that is fired either while reading a request (but only if there are no requests currently queued or running) or later (if the last queued request is finished and there is an incomplete request that expects a continue). That way, we can continue to keep the socket readable and are able to detect client disconnects even if we are in the middle of a expect-continue.

This takes advantage of the fact that a server may omit the 100-continue response if it has already read some of the body of the corresponding request (I read this in the specs).

I am not sure why `send_continue` marks the request as finished - it already did that before, but I do not think that this is correct.

The tests ran fine on my system, but it seems one test failed on github. I guess I will have to take another look at that.

bertjwregeer commented:
> This takes advantage of the fact that a server may omit the 100-continue response if it has already read some of the body of the corresponding request (I read this in the specs).

Where did you read this?

The client hasn't sent a body at this point, and is waiting for the server to respond. Clients implement a fallback that will send the body anyway (sometimes) if no response comes back, but it is absolutely expected that a server sends a 100 continue.

viktordick commented:
> Where did you read this?

I read it [here](https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html):

> An origin server MAY omit a 100 (Continue) response if it has already received some or all of the request body for the corresponding request.

Not sure though if this also holds for HTTP/1.0

viktordick commented:
Alright I was confused regarding the line `request.completed = False`, somehow I mistook it for saying `request.completed = True`. It was somewhat late...
Anyhow, I added a test that a request that uses 100-continue actually receives the body of the request.
I also read [here](https://www.w3.org/Protocols/HTTP/1.0/draft-ietf-http-spec.html#Message-Types) that HTTP/1.0 does not support 100-continue (actually none of the 1xx status code are yet defined in 1.0), so I think this is now ready for a review.

@bertjwregeer I hope I have read the specification correctly regarding omitting of the 100-continue response if a body was already transmitted. If you know of some problem in any browser or proxy or other client with this implementation, we should take that into account, but without any concrete problem I guess we should be safe by relying on the specs.

Anyway, there seems to be one flaky test, namely `test_request_body_too_large_with_no_cl_http10` in `test_functional.py` under `pypy3`. Sometimes the exception `ConnectionClosed` is not raised. But I am not sure if that is a new problem, I think I also saw it once on the master branch.

bertjwregeer commented:
Okay, looking at https://tools.ietf.org/html/rfc7231#section-5.1.1 it does look like the server can ignore sending the 100 Continue if a request body has already been sent by the client, but the standard does say that a client will wait before sending a request body if it has sent an Expect 100 continue request in the first place.

So this may work for the second or third request in a pipelined setup, but we should endeavour to send a reply back as soon as possible.

viktordick commented:
> So this may work for the second or third request in a pipelined setup, but we should endeavour to send a reply back as soon as possible.

Yes, that is what the code does now. If there are no other request currently being executed or queued to be executed, the 100-continue is sent as soon as the request that expects it finished submitting the header (as before). Only if it is a request that is pipelined after other requests which are still being executed, the 100-continue is delayed until all tasks before it have finished, which is necessary to keep the order of responses correct.

So the 100-continue is always written as soon as it is possible to write it without messing up the order of the responses.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -342,10 +374,21 @@ def service(self):
+
+                with self.task_env_lock:
+                    if self.client_disconnected:
+                        self.close_when_flushed = True
```
I think we're good here, the extra lock was removed.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -153,58 +162,70 @@ def handle_read(self):
+        self.current_outbuf_count += num_bytes
+        self.total_outbufs_len += num_bytes
+        self.sent_continue = True
+        self._flush_some()
```
I think this needs to use `_flush_some_if_lockable()` now that it may be invoked from the service thread, not just from the asyncore thread.

mmerickel replied:
The reason I think this is that the asyncore thread could still be writing outbufs even though there is no active request, so you have no guarantee right now that there's no writing going on when you invoke `_flush_some` in `send_continue` from the service thread.

mmerickel replied:
I am similarly concerned with appending data to outbufs without a lock. I think this entire block should be wrapped in `with self.outbuf_lock:` to avoid interfering with an active flush.

mmerickel replied:
Sorry for so many comments here. My proposal is to keep using _flush_some but to wrap the block in a lock:

```python
with self.outbuf_lock:
    self.outbufs[-1].append(outbuf_payload)
    self.current_outbuf_count += num_bytes
    self.total_outbufs_len += num_bytes
    self.sent_continue = True
    self._flush_some()
self.request.completed = False
```

viktordick replied:
If no task is running, only the main thread might call it from `received` and if the main thread is currently processing `received`, it can not simultaneously be in any write operation, can it?

If on the other hand the task thread has written some response that is still in the buffer and then, upon finishing, decides to send the 100-continue of the next request on the channel, which is waiting for it, this might collide with `handle_write` of the main thread. So I guess the suggested change is really necessary.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -360,88 +381,101 @@ def _flush_outbufs_below_high_watermark(self):
+                    and self.request.expect_continue
+                    and self.request.headers_finished
+                    and not self.sent_continue
+                ):
```
`self.request` is shared between the asyncore thread and this thread but not protected by any particular lock. The `task_lock` is acquired here which is great, so It seems like some more effort needs to be made to do something inside `channel.received()` to protect the usage there.

viktordick replied:
I guess you are right.
I used the `task_lock` for all operations on `self.requests` and now also `self.request`. My question is: should I rename it to `requests_lock`? The original purpose of the lock was to protect multiple calls to `service` from actually running, but I guess this changed now.

bertjwregeer replied:
We need to make sure that `channel.received()` does NOT block for extended periods of time. I haven't gone through the code with a fine-tooth comb yet, but please be aware that if we block in `channel.received()` we are blocking inside the main thread... how long and how often do we hold the locks? The original lock was held for the duration of request processing, if that is no longer the case then I think it may be fine.

I am all for renaming variables to match their intent, so I am a +1 on that!

mmerickel replied:
I talked about this with @bertjwregeer offline but I'm quite certain that the original `task_lock` is completely unnecessary in waitress right now because I don't think we ever scheduled the channel concurrently - which is what it was protecting against - ensuring that the channel was not handled by multiple worker threads. With that in mind renaming it totally makes sense.

It's a great point about the locking concern within received. In the previous version of the code received was never invoked while processing a request and now it may be. @viktordick has drastically reduced the time that the lock is held versus what's done in master such that it's not wrapping the entire task lifecycle anymore. There isn't much that really concerns me with how it's being handled right now - it's only really held in 3 places and two of them are around send_continue, with the final being the close_on_finish teardown.

The send_continue is potentially long running but I do not see how it's any worse than on master.

mmerickel commented on src/waitress/channel.py:
```diff
@@ -153,58 +163,70 @@ def handle_read(self):
-                request = None
-            else:
-                self.request = request
+            with self.task_lock:
```
I'd just move this lock around the entire while loop instead of releasing it each iteration. It seems unnecessary to me but open to arguments otherwise.

viktordick replied:
If the client sends a really large request that needs several iterations to be parsed, a running task that is finished might be able to queue the next task (which requires removing the first element of `self.requests` and therefore acquiring the lock) between two loop iterations if the lock is released. But I am not sure if this is a realistic scenario, if the actual processing of received data is usually very quick it might not be worth the overhead of releasing and re-acquiring the lock.

mmerickel replied:
Yeah my feeling is mainly that the data has already been read so the loop should just go as fast as possible. The only reason not to hold the lock is if expect-continue was frequent - which it's not.

mmerickel commented:
@viktordick this is awesome, in my mind the only thing missing is a changelog entry!

viktordick commented:
@mmerickel Thanks! I wrote a changelog entry and squashed the commits.

@bertjwregeer Are you also happy with the result? There is still a change request from you recorded on the PR, but I think everything mentioned there has been addressed in the mean time.

bertjwregeer commented:
Thanks for being patient and working through this. I want to read this critically one last time and make sure that I've convinced myself that there are no locking issues/race conditions.

Thanks for all your hard work!

bertjwregeer commented:
Thanks so much @viktordick for your patience and working through this. This looks great, and I have adequately convinced myself that this won't break!

viktordick commented:
Thanks a lot for the help and patience!

Is a release already planned sometime soon?

bertjwregeer commented:
@viktordick https://pypi.org/project/waitress/2.0.0b0/

I need testers to validate this doesn't break anything, but there is now a released version of waitress you can try in your environment.

viktordick commented:
Thanks.
We are actually already using it - we simply use

    git+https://github.com/perfact/waitress@2.0.0dev0.perfact.2

In our `requirements.txt`, which is essentially the current state of upstream `master` plus a commit that gives a custom tag. Although the number of systems that actually already use it is still small (I think no production system yet).
